collect/cache/span_cache_test.go

```diff
@@ -1,106 +1,285 @@
+
+			ids := make([]string, b.N)
+			for i := 0; i < b.N; i++ {
+				ids[i] = genID(32)
+			}
+			evt := types.Event{
+				APIHost: "apihost",
+				APIKey:  "apikey",
+				Dataset: "dataset",
+			}
 
-	ids := make([]string, b.N)
-	for i := 0; i < b.N; i++ {
-		ids[i] = genID(32)
+			for i := 0; i < b.N; i++ {
+				span := &types.Span{
+					TraceID: ids[i],
+					Event:   evt,
+				}
+				c.Set(span)
+			}
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				c.Get(ids[i])
+			}
+		})
 	}
-	evt := types.Event{
-		APIHost: "apihost",
-		APIKey:  "apikey",
-		Dataset: "dataset",
+}
+
+func BenchmarkSpanCacheGetTraceIDs(b *testing.B) {
+	for _, typ := range []string{"basic", "complex"} {
+		c := getCache(typ, clockwork.NewFakeClock())
+		b.Run(typ, func(b *testing.B) {
+
+			c.(startstop.Starter).Start()
+			defer c.(startstop.Stopper).Stop()
+
+			ids := make([]string, b.N)
+			for i := 0; i < b.N; i++ {
+				ids[i] = genID(32)
+			}
+			evt := types.Event{
+				APIHost: "apihost",
+				APIKey:  "apikey",
+				Dataset: "dataset",
+			}
+
+			for i := 0; i < b.N; i++ {
+				span := &types.Span{
+					TraceID: ids[i],
+					Event:   evt,
+				}
+				c.Set(span)
+			}
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				c.GetTraceIDs(50)
+			}
+		})
 	}
+}
 
-	for i := 0; i < b.N; i++ {
-		span := &types.Span{
-			TraceID: "",
-			Event:   evt,
-		}
-		c.Set(span)
+func BenchmarkSpanCacheGetOldest(b *testing.B) {
+	for _, typ := range []string{"basic", "complex"} {
+		c := getCache(typ, clockwork.NewRealClock())
+		b.Run(typ, func(b *testing.B) {
+
+			c.(startstop.Starter).Start()
+			defer c.(startstop.Stopper).Stop()
+
+			ids := make([]string, b.N)
+			for i := 0; i < b.N; i++ {
+				ids[i] = genID(32)
+			}
+			evt := types.Event{
+				APIHost: "apihost",
+				APIKey:  "apikey",
+				Dataset: "dataset",
+			}
+
+			for i := 0; i < b.N; i++ {
+				span := &types.Span{
+					TraceID:     ids[i],
+					Event:       evt,
+					ArrivalTime: c.GetClock().Now(),
+				}
+				c.Set(span)
+			}
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				c.GetOldest(.1)
+			}
+		})
 	}
-	b.ResetTimer()
+}
+
+func BenchmarkSpanCacheMixed(b *testing.B) {
+	const numIDs = 10000
+	for _, typ := range []string{"basic", "complex"} {
+		c := getCache(typ, clockwork.NewFakeClock())
+		b.Run(typ, func(b *testing.B) {
+
+			c.(startstop.Starter).Start()
+			defer c.(startstop.Stopper).Stop()
+
+			ids := make([]string, numIDs)
+			for i := 0; i < numIDs; i++ {
+				ids[i] = genID(32)
+			}
+			evt := types.Event{
+				APIHost: "apihost",
+				APIKey:  "apikey",
+				Dataset: "dataset",
+			}
 
-	for i := 0; i < b.N; i++ {
-		c.Get(ids[i])
+			b.ResetTimer()
+			// we have numIDs IDs, and we'll iterate and for each
+			// ID we'll either:
+			// - add a span (80% of the time)
+			// - get a trace (10% of the time)
+			// - delete a trace (10% of the time)
+			for i := 0; i < b.N; i++ {
+				switch rand.Intn(10) {
+				case 0:
+					c.Remove(ids[i%numIDs])
```

Comment:
Should we put each call into a goroutine so we can benchmark the mix with contention?

Reply:
In my view, that's a different benchmark -- and I haven't come up with a good way to write that one yet.